docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,562 @@
+   This allows for some housekeeping of the `vmetadata` bucket (see below)
+
+
+### Metadata housekeeping
+
+**Nice to have**. A background job periodically cleans up metadata for all terminated diamonds and their splits.
+
+1. Regular users and service accounts are not authorized to delete objects in buckets.
+2. After a commit or cancel has been carried out, we may safely remove all diamond-related metata.
+3. We don't want metadata buckets to grow and hold forever what is essentially temporary.
+4. A privileged job is in responsible for cleaning, for all repositories in a given context, all references to terminated
+   diamonds and their splits in the vmetadata bucket.
+
+> **NOTE**: while not critical, this task guards our `vmetadata` bucket against undesirable pollution from tracks of
+> failed CI jobs, etc. We could run this on every week's end for instance.
+
+
+## Example
+
+```bash
+repo="test-repo"
+diamond=$(datamon diamond initialize --repo "${repo}")
+
+datamon diamond add --repo "${repo}" --diamond "${diamond-id}" --path /datasource/part1 &
+datamon diamond add --repo "${repo}" --diamond "${diamond-id}" --path /datasource/part2 &
+datamon diamond add --repo "${repo}" --diamond "${diamond-id}" --path /datasource/part3 &
+
+datamon diamond split list --repo "${repo}" --diamond "${diamond-id}"
+
+datamon diamond commit --repo "${repo}" --diamond "${diamond-id}" --message "Diamond constructed"
+```
+
+
+## Detailed design
+
+### Diamonds vs incremental uploads
+
+Diamonds and incremental uploads are very much alike, the use-cases come with some nuances.
+
+Use-cases:
+
+1. Diamond: a workflow spawns many pods to produce a single bundle as output.
+2. Incremental uploads from a single client: a single pod eventually produces a bundle, but proceeds in several steps,
+   with intermediate checkpoints, to persist the output.
+3. Incremental uploads in parallel: many pods output data incrementally, and eventually produce a bundle.
+   We don't want conflicts between increments but would like to track conflicts between nodes.
+
+The main difference lies with concurrency and requirements on client coordination: a diamond carried out by a single
+client _is_ an incremental upload. The other difference is how the user wants to deal with conflicts.
+
+Therefore, we need to capture user intent. We add some additional CLI flags to specify how datamon should deal
+with conflicts. See also [§ Handling Conflicts below](#handling-conflicts).
+
+
+Available options at commit time:
+```
+datamon diamond commit --diamond {diamond-id} \
+                      [--with-conflicts]   # <- this is the default - overlaps are reported as conflicts detected when merging splits
+
+datamon diamond commit --diamond {diamond-id} \
+                       --with-checkpoints   # <- indicates an incremental upload: conflicts are subsequent versions of uploaded files
```

Comment:
I still prefer the client reusing the same split-id and letting the most recent file win. I don't yet see the reason to have this flag.